lib/pacemaker/pcmk_graph_producer.c

```diff
@@ -182,119 +182,131 @@ add_downed_nodes(xmlNode *xml, const pe_action_t *action,
-    if (action->rsc != NULL && action->rsc->clone_name != NULL) {
+
+    if ((action->rsc != NULL) && (action->rsc->clone_name != NULL)) {
         char *clone_key = NULL;
         guint interval_ms;
 
-        if (pcmk__guint_from_hash(action->meta,
-                                  XML_LRM_ATTR_INTERVAL_MS, 0,
+        if (pcmk__guint_from_hash(action->meta, XML_LRM_ATTR_INTERVAL_MS, 0,
                                   &interval_ms) != pcmk_rc_ok) {
             interval_ms = 0;
         }
-
-        if (pcmk__str_eq(action->task, RSC_NOTIFY, pcmk__str_casei)) {
+        if (pcmk__str_eq(action->task, RSC_NOTIFY, pcmk__str_none)) {
             const char *n_type = g_hash_table_lookup(action->meta, "notify_type");
-            const char *n_task = g_hash_table_lookup(action->meta, "notify_operation");
+            const char *n_task = g_hash_table_lookup(action->meta,
+                                                     "notify_operation");
 
-            CRM_CHECK(n_type != NULL, crm_err("No notify type value found for %s", action->uuid));
-            CRM_CHECK(n_task != NULL,
-                      crm_err("No notify operation value found for %s", action->uuid));
+            CRM_LOG_ASSERT((n_type != NULL) && (n_task != NULL));
             clone_key = pcmk__notify_key(action->rsc->clone_name,
                                          n_type, n_task);
 
-        } else if(action->cancel_task) {
+        } else if (action->cancel_task != NULL) {
             clone_key = pcmk__op_key(action->rsc->clone_name,
                                      action->cancel_task, interval_ms);
         } else {
             clone_key = pcmk__op_key(action->rsc->clone_name,
                                      action->task, interval_ms);
         }
-
-        CRM_CHECK(clone_key != NULL, crm_err("Could not generate a key for %s", action->uuid));
         crm_xml_add(action_xml, XML_LRM_ATTR_TASK_KEY, clone_key);
         crm_xml_add(action_xml, "internal_" XML_LRM_ATTR_TASK_KEY, action->uuid);
         free(clone_key);
-
     } else {
         crm_xml_add(action_xml, XML_LRM_ATTR_TASK_KEY, action->uuid);
     }
 
-    if (needs_node_info && action->node != NULL) {
+    if (needs_node_info && (action->node != NULL)) {
         pe_node_t *router_node = pcmk__connection_host_for_action(action);
 
-        crm_xml_add(action_xml, XML_LRM_ATTR_TARGET, action->node->details->uname);
-        crm_xml_add(action_xml, XML_LRM_ATTR_TARGET_UUID, action->node->details->id);
-        if (router_node) {
-            crm_xml_add(action_xml, XML_LRM_ATTR_ROUTER_NODE, router_node->details->uname);
+        crm_xml_add(action_xml, XML_LRM_ATTR_TARGET,
+                    action->node->details->uname);
+        crm_xml_add(action_xml, XML_LRM_ATTR_TARGET_UUID,
+                    action->node->details->id);
+        if (router_node != NULL) {
+            crm_xml_add(action_xml, XML_LRM_ATTR_ROUTER_NODE,
+                        router_node->details->uname);
```

Comment:
It wouldn't really save much (in fact it's implemented as crm_xml_add internally) but you could use pcmk__xe_set_propv to set the first two in one go.

Reply:
not a bad idea but probably not worth changing for just 2